add deprecate guidelines doc [draft] (#150)
nodejs/package-maintenance: docs/drafts/DEPRECATE-GUIDELINES.md

# What?

This is guidelines for unmaintained packages.

# Why?

To ensure that packages in the Node.js ecosystem are up to date or gracefully deprecated.

# How?

> These guidelines exist to help with the course of actions for package consumers and owners to take in the event that a package becomes unmaintained.

Reviewer: ```suggestion
These guidelines exist to help with a course of action for package consumers and owners to take in the event that a package becomes unmaintained.
```

- Provide a baseline and common practices for identifying unmaintained packages
- Give package consumers some help with unmaintained packages.
- Give package owners some help with handling and deprecating packages.

## What consititutes an unmaintained package?

> It's not always obvious if a package is unmaintained. Many packages are small and considered completed as published. Long period of no activity does not mean unmaintain.

Reviewer: ```suggestion
It's not always obvious if a package is unmaintained. Many packages are small and considered completed as published. A long period of no activity does not imply that a project is not maintained.
```

> Generally the question "is this package unmaintained?" arises when users found it to have issues that are not getting reponse or being addressed.

Reviewer: ```suggestion
Generally the question "is this package unmaintained?" arises when users find it to have issues that are not getting responded to, or being addressed.
```

> Some criterias to determine if it's unmaintained:

Reviewer: ```suggestion
Some criteria to determine if it's unmaintained:
```

> - The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there inquries" for critical issues.

Reviewer: ```suggestion
- The author is no longer responding to questions, issues, or PRs; making any updates; and specifically, not responding to repeat "are you there" inquiries for critical issues.
```

- The author may have explicitly indicated that they will stop all activities on the package.

> - A different package that's more active and the author acknowledged as the replacement.

Reviewer: ```suggestion
- The author has explicitly indicated that an alternative package should be used instead.
```

Reviewer 2: @jchip  can you accept this suggestion?

Reviewer: ping @jchip

> - Critical issues exist for the package and not being addressed

Reviewer: ```suggestion
- Critical issues exist for the package that are not being addressed
```

Reviewer 2: @jchip can you accept this suggestion

Reviewer: ping here too

  - Known vulnerabilities identified by `npm audit` or other parties
  - Package is known to fail for LTS Node.js

## For Users

If you are using a package that meets your requirements but you found issues or need enhancements, then these are some actions you can follow:

- Note the deprecation messages during `npm install` and follow the instruction if the message offered any.

> - Check the package's npm page at `https://www.npmjs.com/package/<package-name>`. ie: <https://www.npmjs.com/package/express>

Reviewer 3: I would add also "Check the eventually `support` field in package.json" (refer to pr #139 )

Then I think this doc is clear enough ✌

>   - Check the "last publish" date and versions history to get a sense of its latest activies.

Reviewer: ```suggestion
  - Check the "last publish" date and versions history to get a sense of its latest activity. Note that you may want to `npm show $package time` to see all the published versions and times, regardless of their dist-tag setting.
```

Reviewer: and here

>   - Open the "repository" link if it has one to view development activities.

Reviewer: ```suggestion
  - Open the "repository" link if it has one to view development activities (you can run `npm repo $package` to do this as well)
```

### Identifying unmaintained package?

When you've opened the package's repo, to further check development activities:

> - Check README or issues to see if the author indicated they are no longer interested in maintaining the package

Reviewer: ```suggestion
- Check README or issues to see if the author has indicated they are no longer interested in maintaining the package
```

> - Search through existing issues and PRs to see if anyone else has submitted inquries similar to what you are seeking.

Reviewer: ```suggestion
- Search through existing issues and PRs (closed and open) to see if anyone else has submitted inquiries similar to what you are seeking.
```

> - Check for or file issue in the repo to encourage author to deprecate a version that:

Reviewer: ```suggestion
- Check for or file an issue in the repo to encourage the author to replace and deprecate a version that:
```

  - Has known critical bugs and should be avoided

>   - Known to fail for LTS Node.js

Reviewer: ```suggestion
  - Is known to fail for LTS node.js
```

- Check last activity on github issues and PRs. If there're no responses within some reasonable time to high priority issues or PRs, then consider the package as an unmaintained candidate.

### Taking Further Actions

Once you've identified a package may be unmaintained, please open an issue for the Node.js package-maintenance working group to check and identify path forward.

## For Package Owners

If you own a package and you want to stop maintaining it, then please help your users by making it known. Some simple actions you can take:

- Use the `npm deprecate` command to mark package as unmaintained.
- Add a note to your repo's `README.md` file.
- Reach out to the Node.js package-maintenance working group to offer us ownership so we can hand it off to a new owner.

### `npm deprecate` command

- npm has the "deprecate" command, you can utilize this.
- deprecate your latest version with a message that includes `"abandoned and unmaintained"`.
  - It allows updating the message so you can change it later - verified as of 02/05/2019
  - It can be undone by setting an empty message.
    - ie: `npm deprecate package@1.0.0 ""`

> - Even if you are still maintaining the package, please consider marking an old version as deprecated.

Reviewer: I'm not so sure about this one - I deprecate old versions when they have critical bugs or vulns, but in general, using old versions is totally fine forever, and adding noise to these users' installs doesn't help anything.

Author: update to:

Author: - Please consider marking an old version with crtical issues or vulnerabilities as deprecated.

### Mark abandoned/unmaintained

- Use a special `npm deprecate` message such as `"abandoned and unmaintained"` to mark a package as abandoned.

  - no need to publish a new package to mark it abandoned.
  - but only owners can `npm deprecate` a package.
  - please consider giving the package-maintenance working group ownership to the package on npm.

### Identify replacement

- If a package is fully unmaintained, then a replacement should be identified and added to the deprecation message.
- If no replacement exists, then please identify the safe versions to use in the deprecation message.

## Further Ideas and Tools

There are some ongoing discussion related to npm to make managing unmaintained packages easier.

- Package override RFC https://github.com/aeschright/rfcs/blob/ca127d7c5fb7ecde534172d771351c0b8f819e79/accepted/0009-package-overrides.md

Other ideas for tools that could help:

> - A cli to allow `npm deprecate` a range of versions on a package.

Reviewer: This is already possible with npm; `npm deprecate $package@'^1||^2' $message`, say.
```suggestion
```

- or request new support of `npm deprecate` to allow deprecating versions published before a given time

> - allow consumer to specify a specific version to use for a package through an `override` field in their package.json

Reviewer: ```suggestion
 - use npm 6.9+'s aliasing ability to provide a replacement for transitive deps.
```

  - if package fails to install/build for an LTS release of node.js

- Check last activity on github issues and PRs. If there're no responses within some reasonable time to high priority issues or PRs, then consider the package as an unmaintained candidate.

### Taking Further Actions

Once you've identified a package may be unmaintained, there are some basic actions you can take to help yourself:

- Fork the repo and enhance the code yourself.
  - If you don't want to publish a modified copy under a different name, then consider using a tool like <https://github.com/ds300/patch-package>.
- Consider becoming a maintainer for the package.
- Search for a maintained alternative that may meet your requirements better.
- Please open an issue for the Node.js package-maintenance working group so that we may request for npm ownership in case someone's interested in maintaining the package.

## For Package Owners

If you own a package and you want to stop maintaining it, then please help your users by making it known. Some simple actions you can take:

- Use the `npm deprecate` command to mark package as unmaintained.
- Add a note to your repo's `README.md` file.

> - Reach out to the Node.js package-maintenance working group to offer us ownership so we can hand it off to a new owner.

Reviewer 4: Should this include updating the `support` levels in the `package.json`? That would require republishing the package, but maybe there's a way to make that easier, somehow built into the `deprecate` command?

# What?

Guidelines for identifying and marking unmaintained packages.

# Why?

Provide a baseline and common practices for handling packages that are unmaintained.

# How?

These guidelines exist to help with the course of actions to take in the event that a package becomes unmaintained.

## What consititutes an unmaintained package?

> - The author is no longer responding to questions, issues, PRs, or making any updates.

Reviewer: maybe worth qualifying here that they’re updates needed for particular reasons? if the only things the author is ignoring are feature requests the author doesn’t want, the package isn’t necessarily unmaintained.

- The author may have explicitly indicated that they will stop all activities on the package.
- A different package that's more active and the author acknowledged as the replacement.
- Critical issues exist for the package and not being addressed
  - Known vulnerabilities identified by `npm audit` or other parties

>   - Package is known to fail for LTS NodeJS

Reviewer: ```suggestion
  - Package is known to fail for LTS node.js
  - semver violations (any non-bug fix break in a non-major, eg)
```

## Identifying unmaintained package?

- check last activity on github issues and PRs.

  - mark package as unmaintained candidate if no response from author to issues or PRs within one year

- file high priority issues in the package's repo

  - if npm audit identified vulnerabilities that are critical

>   - if package is broken or fail to install for a LTS release of NodeJS

Reviewer: ```suggestion
  - if fails to install/build for an LTS release of node.js
```
what constitutes “broken”?

  - if author indicate they are no longer interested in maintaining, then add as unmaintained candidate
  - if no response within three months, then add as unmaintained candidate.

## `npm deprecate` command

- npm already has the "deprecate" command, we should utilize this.

  - It allows updating the message - verified as of 02/05/2019
  - It can be undone by setting an empty message. ie: `npm deprecate package@1.0.0 ""`

  - It can be used to mark a version as deprecate while the package may be very active still.

## Mark abandoned/unmaintained

- Use a special `npm deprecate` message such as `"abandoned and unmaintained"` to mark a package as abandoned.

  - no need to publish a new package to mark it abandoned.
  - but only owners can `npm deprecate` a package.
  - if author simply can't be reached and package is very outdated, then need to contact npm to get access to deprecate package.

> * A cli to allow `npm deprecate` a range of versions on a package.

Reviewer: that’s already how `npm deprecate` works

> * or `npm deprecate` versions with published date older than a given time

Reviewer: that seems an interesting feature to propose to npm, but it’s out of scope of this group

## Identify replacement

> - If a package is fully unmaintained, then a replacement should be identify and add to the deprecate message.

Reviewer: ```suggestion
- If a package is fully unmaintained, then a replacement should be identified and added to the deprecation message.
```

> - If no replacement exist, then should identify the safe versions to use in deprecate message.

Reviewer: ```suggestion
- If no replacement exists, then should identify the safe versions to use in the deprecation message.
```

## Encourage Author to deprecate versions

- File issue in repo to encourage author to deprecate a version that:
  - Has known critical bugs and should be avoided

>   - Known to fail for LTS NodeJS

Reviewer: ```suggestion
  - Known to fail for LTS node.js
```
(not sure what NodeJS is :-p)

  - Has known critical vulnerabilities

> 

Reviewer 3: Could we add "Suggest to join the package-maintenance program"?
Some package could need help and doesn't know we are here to help them (with tools almost)

  - Check the "last publish" date and versions history to get a sense of its latest activies.
  - Open the "repository" link if it has one to view development activities.

### Identifying unmaintained package?

When you've opened the package's repo, to further check development activities:

- Check README or issues to see if the author indicated they are no longer interested in maintaining the package

- Search through existing issues and PRs to see if anyone else has submitted inquries similar to what you are seeking.

- Check for or file issue in the repo to encourage author to deprecate a version that:

  - Has known critical bugs and should be avoided
  - Known to fail for LTS Node.js
  - Has known critical vulnerabilities

- Check if high priority issues in the package's repo are getting responses from owner

  - if npm audit identified vulnerabilities that are critical
  - if package fails to install/build for an LTS release of node.js

- Check last activity on github issues and PRs. If there're no responses within some reasonable time to high priority issues or PRs, then consider the package as an unmaintained candidate.

### Taking Further Actions

> Once you've identified a package may be unmaintained, please open an issue for the Node.js package-maintenance working group to check and identify path forward.

Reviewer 4: Is this the right recommendation? I would like to see package-maintenance group to be the last resort (depending on popularity of the package, of course), but I would generally recommend looking for alternatives (are there any tools out there?) and then forking, i.e. dyi?

Reviewer: Forking is an important ability, but I see that as more of a last resort than asking this group - creating forks further fractures the ecosystem, and doesn't help all of the users who are still using the unmaintained version unwittingly.

Author: if no one can reach the author then there's really not much anyone could do.  I was thinking that the best this group could do is get npm ownership and actively advertise for someone interested in becoming the maintainer, and/or add deprecate message to identify alternative if one exist.

### `npm deprecate` command

- npm has the "deprecate" command, you can utilize this.
- deprecate your latest version with a message that includes `"abandoned and unmaintained"`.
  - It allows updating the message so you can change it later - verified as of 02/05/2019
  - It can be undone by setting an empty message.
    - ie: `npm deprecate package@1.0.0 ""`
- Even if you are still maintaining the package, please consider marking an old version as deprecated.

### Mark abandoned/unmaintained

- Use a special `npm deprecate` message such as `"abandoned and unmaintained"` to mark a package as abandoned.

  - no need to publish a new package to mark it abandoned.
  - but only owners can `npm deprecate` a package.
  - please consider giving the package-maintenance working group ownership to the package on npm.

### Identify replacement

- If a package is fully unmaintained, then a replacement should be identified and added to the deprecation message.
- If no replacement exists, then please identify the safe versions to use in the deprecation message.

## Further Ideas and Tools

There are some ongoing discussion related to npm to make managing unmaintained packages easier.

> - Package override RFC https://github.com/aeschright/rfcs/blob/ca127d7c5fb7ecde534172d771351c0b8f819e79/accepted/0009-package-overrides.md

Reviewer 4: [npm@6.9.0](https://npm.community/t/release-npm-6-9-0/5911) allows package aliases - I have not yet looked into it in depth, and while the linked RFC here covers a method for override discovery, once a suitable replacement is identified - it might be worth mentioning the alias functionality?

## What consititutes an unmaintained package?

> - The author is no longer responding to questions, issues, PRs, or making any updates, and specifically repeat "are you there" inquries for critical issues.

Reviewer 2: `and specifically repeat "are you there" inquries for critical issues.`  maybe this should be

` and has not responded to repeated "are you there inquiries" on critical issues`

### Taking Further Actions

Once you've identified a package may be unmaintained, there are some basic actions you can take to help yourself:

- Fork the repo and enhance the code yourself.
  - If you don't want to publish a modified copy under a different name, then consider using a tool like <https://github.com/ds300/patch-package>.

> - Consider becoming a maintainer for the package.

Reviewer 5: How can somebody do so? we need to add some explanation/link here. https://docs.npmjs.com/misc/disputes.html is the source of truth.

Author: The npm disputes guideline looks to be mainly for package name conflict or squatting resolution.

For a random user from the internet to take over a package, we need to establish their intent else we risk the same thing as the event-stream incident.

Then we need to establish how to gain ownership.  If the owner can at least be reached and gladly hand off ownership then that's easy.  Otherwise one could potentially use the disputes process.

I think this WG could help facilitate the process by maintaining a doc of packages that's pending seeking new maintainer and contact with the original owner.

That's why this WG would serve to help with getting ownership parking to eventually hand it off.

I think the part where "how can we establish this user's intent and let them take over this package" is one of the key things that triggered some of the discussions.

Is establishing some steps for vetting a user to take ownership part of this WG's purpose?

Reviewer 4: I think linking to the disputes doc is a good idea - it does cover the case of unmaintained packages and it does provide the steps and this group would probably have to use that process anyways - the decision to hand over the package will still fall onto npm, so even if the request goes via this WG (this should still be the last resort), the link is handy.

> Is establishing some steps for vetting a user to take ownership part of this WG's purpose?

 I think we discussed that it is. Covered in https://github.com/nodejs/package-maintenance/issues/155?
